src/scheduler/pbs_sched.c

```diff
@@ -1273,3 +1193,127 @@ main(int argc, char *argv[])
+ * @brief
+ *		Compares two sched commands and tells whether they are duplicate or not.
+ *
+ * @param[in]	cmd1	-	command1
+ * @param[in]	cmd2	-      command2
+ *
+ *
+ * @return	int
+ * @retval	1	-	if both commands are duplicate
+ * @retval	0	-	if not duplicate
+ *
+ */
+static int
+compare_sched_cmd(const sched_cmd *cmd1, const sched_cmd *cmd2)
+{
+	if (cmd1 && cmd2) {
+		if (cmd1->jid == NULL && cmd2->jid == NULL) {
+			if (cmd1->cmd == cmd2->cmd)
+				return 1;
+		} else {
+			if (cmd1->cmd == cmd2->cmd && !strcmp(cmd1->jid, cmd2->jid))
+				return 1;
+		}
+	}
+	return 0;
+}
+
+/**
+ * @brief Filters scheduler commands queue and discards duplicates
+ *
+* @param[in] sched_cmds	- scheduler command queue
+ *
+ * @return	ds_queue *
+ * @retval	success	- returns the filtered queue
+ * @reval	 fail -	NULL
+ *
+ */
+ds_queue *
+filter_sched_cmds(ds_queue *sched_cmds)
+{
+	/* sched_cmds_filtered is a  resultant queue which consists of sched commands after discarding
+	 * the duplicate commands which is finally returned from this function.
+	 */
+	ds_queue *sched_cmds_filtered = new_ds_queue();
+	sched_cmd *tmp_cmd;
+	int cmds_front;
+	int cmds_rear;
+	sched_cmd *cmd_enq;
+
+	if (sched_cmds_filtered == NULL) 
+		goto error;
+
+	/* 1. Traverse sched_cmds queue
+	 * 2. While we traverse we keep pushing the unique elements into the sched_cmds_filtered queue
+	 * 3. To find out if a command is duplicate or not here we are making use of the function compare_sched_cmd
+	 * 4. Finally return the sched_cmds_filtered_queue which now only consists of unique sched commands.
+	 */
+
+	cmds_front = sched_cmds->front;
+	cmds_rear = sched_cmds->rear;
+
+
+	tmp_cmd = sched_cmds->content_arr[cmds_front];
+	if (tmp_cmd == NULL)
+		goto error;
+
+	cmd_enq = new_sched_cmd();
+	if (cmd_enq == NULL)
+		goto error;
+
+	cmd_enq->cmd = tmp_cmd->cmd;
+	cmd_enq->from_sock = tmp_cmd->from_sock;
+	if (tmp_cmd->jid != NULL) {
+		cmd_enq->jid = string_dup(tmp_cmd->jid);
+		if (cmd_enq->jid == NULL)
+			goto error;
+	}
+
+	if (!ds_enqueue(sched_cmds_filtered, cmd_enq)) 
+		goto error;
+	cmds_front++;
+
+	for (; cmds_front < cmds_rear; cmds_front++) {
+		int no_enq = 0;
+		int cmds_fltrd_front = sched_cmds_filtered->front;
+		int cmds_fltrd_rear = sched_cmds_filtered->rear;
+		for (; cmds_fltrd_front < cmds_fltrd_rear; cmds_fltrd_front++) {
+			if (compare_sched_cmd(sched_cmds_filtered->content_arr[cmds_fltrd_front],
+				sched_cmds->content_arr[cmds_front]) == 1) {
+					no_enq = 1;
+					break;
+			} 
+		}
+		if (no_enq == 0) {
+			sched_cmd *cmd_enq;
+			sched_cmd *tmp_cmd;
+			cmd_enq = new_sched_cmd();
+			if (cmd_enq == NULL)
+				goto error;
+			tmp_cmd = sched_cmds->content_arr[cmds_front];
+
+			cmd_enq->cmd = tmp_cmd->cmd;
+			cmd_enq->from_sock = tmp_cmd->from_sock;
+			if (tmp_cmd->jid != NULL) {
+				cmd_enq->jid = string_dup(tmp_cmd->jid);
+				if (cmd_enq->jid == NULL)
+					goto error;
+			}
+
+			if (!ds_enqueue(sched_cmds_filtered, cmd_enq)) 
+				goto error;		
```

Comment:
I don't think a queue is the right data structure for what you want to do now.  Before it was fine.  Once you start making your queue a hybrid, you should rethink your data structure.  What I think you should do is have 2 data structures.  One is just an array which is the size of the number of commands (17?).  When a new command comes in, you set array[cmd] = 1.  If 20 of the same command come in, you'll just set the same element 20 times.  When you need to run commands, run down the array, and run a cycle for every index which is set to 1.  The second data structure is a list of qruns.  I'd run the qruns first, and then the array of normal cycles.

Reply:
Thanks for suggesting these changes @bhroam. Apart from this I had to do some more changes. Summary of the changes is given below.

- Instead of storing the commands in a queue now I am storing them in a hash table implemented using an array like the way you suggested.
- Storing the list of SCHEDULE_AJOBS commands in a separate list 
- Processing all the qrun_list first followed by the rest of the commands
- End of cycle indication is now sent to all the servers to accommodate the above changes.
Please look into my latest commit for the code changes.